### PR TITLE
MOEA/D

### PR DESCRIPTION
**Pull request by bjornbugge**
Grid functions and MOEA/D algorithm

### CODE REVIEW

**Comment by Ulfgard** [comment]
hej, clang hates your override:

/Users/travis/build/Shark-ML/Shark/include/shark/Algorithms/DirectSearch/MOEAD.h:135:39: error: only virtual member functions can be marked 'override'

    void serialize(Archive & archive) override

                                      ^~~~~~~~

/Users/travis/build/Shark-ML/Shark/include/shark/Algorithms/DirectSearch/MOEAD.h:167:10: warning: 'init' overrides a member function but is not marked 'override' [-Winconsistent-missing-override]

    void init(ObjectiveFunctionType & function, 

         ^

/Users/travis/build/Shark-ML/Shark/include/shark/Algorithms/AbstractMultiObjectiveOptimizer.h:93:15: note: overridden virtual function is here

        virtual void init( 

                     ^

File: include/shark/Algorithms/DirectSearch/MOEAD.h [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+        m_neighbourhoodSize = neighbourhoodSize;
+        m_neighbourhoods = closestIndices(m_weights, 
+                                          neighbourhoodSize);
+        
+        SIZE_CHECK(m_weights.size1() == mu);
+        SIZE_CHECK(m_neighbourhoods.size1() == mu);
+        m_mu = mu;
+        m_mutation.m_nm = nm;
+        m_crossover.m_nc = nc;
+        m_crossoverProbability = crossover_prob;
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/MOEAD.h]: something with the tabs is wrong here...

File: include/shark/Algorithms/DirectSearch/MOEAD.h [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+            m_bestDecomposedValues[i] = std::min(m_bestDecomposedValues[i], 
+                                                 candidate[i]);
+        }
+        // 2.4. Update of neighbouring solutions
+        for(auto iter = m_neighbourhoods.row_begin(m_curParentIndex);
+            iter != m_neighbourhoods.row_end(m_curParentIndex);
+            ++iter)
+        {
+            std::size_t j = *iter;
+            const RealVector lambda_j(m_weights.row_begin(j),
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/MOEAD.h]: lambda_j = row(m_weights,j);

File: include/shark/Algorithms/DirectSearch/MOEAD.h [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+        const IndividualType & offspring = offspringvec[0];
+        // 2.3. Update the "Z" vector.
+        RealVector candidate = offspring.unpenalizedFitness();
+        for(std::size_t i = 0; i < candidate.size(); ++i)
+        {
+            m_bestDecomposedValues[i] = std::min(m_bestDecomposedValues[i], 
+                                                 candidate[i]);
+        }
+        // 2.4. Update of neighbouring solutions
+        for(auto iter = m_neighbourhoods.row_begin(m_curParentIndex);
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/MOEAD.h]: for(std::size_t j: row(m_neighborhoods,m_curParentIndex)){

File: include/shark/Algorithms/DirectSearch/Operators/Grid.h [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+namespace shark {
+
+
+/*
+ * An n-dimensional point sums to 's' if the sum of the parts equal 's',
+ * e.g. the point (x_0,x_1,x_2) sums to x_0+x_1+x+2 etc.  The number of
+ * n-dimensional points that sum to 's' is given by the formula "N over K" where
+ * N is n - 2 + s + 1 and K is s. 
+ */
+std::size_t sumlength(std::size_t const n, std::size_t const sum)
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/Operators/Grid.h]: is that needed? maybe also detail?

File: include/shark/Algorithms/DirectSearch/Operators/Grid.h [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+                  {
+                      return my_dists[a] < my_dists[b];
+                  });
+        // Copy the T closest indices into B.
+        std::copy_n(indices.begin(), t, neighbourIndices.row_begin(i));
+    }
+    return neighbourIndices;
+}
+
+void dump(RealMatrix const & m, std::string const & filename)
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/Operators/Grid.h]: please remove this

File: include/shark/Algorithms/DirectSearch/Operators/Grid.h [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+    UIntMatrix pointMatrix(point_count, n);
+    detail::pointLattice_helper(pointMatrix, 0, 0, sum);
+    return pointMatrix;
+}
+
+
+/*
+ * Sample points uniformly from the simplex given in the matrix.  Corners are
+ * always included in the sampled point set (unless excplicitly turned off with
+ * keep_corners set to false).  The returned matrix will always have n points or
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/Operators/Grid.h]: are points sampled uniquely? maybe add :)

File: include/shark/Algorithms/DirectSearch/Operators/Grid.h [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Shark.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+//===========================================================================
+
+#ifndef SHARK_ALGORITHMS_DIRECT_SEARCH_OPERATORS_GRID
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/Operators/Grid.h]: In general: be a bit careful with generic names in the shark namespace. "closestIndices" is quite generic and can occur in different contexts meaning different things. Maybe encapsulate everything here into a class to get rid of this problem

File: include/shark/Algorithms/DirectSearch/Operators/Scalarizers/Tchebycheff.h [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ */
+//===========================================================================
+
+#ifndef SHARK_ALGORITHMS_DIRECT_SEARCH_OPERATORS_SCALARIZERS_TCHEBYCHEFF
+#define SHARK_ALGORITHMS_DIRECT_SEARCH_OPERATORS_SCALARIZERS_TCHEBYCHEFF
+
+namespace shark {
+
+
+double tchebycheff(RealVector const & fitness,
```
Comment (Ulfgard) [inline include/shark/Algorithms/DirectSearch/Operators/Scalarizers/Tchebycheff.h]: maybe tchebycheffScalarizer?

**Comment by Ulfgard** [comment]
Hi. Please check my changes. I have seen that there were a few whitespace "mishaps" and other small things that i just cleaned up. Most changes are:

naming: I renamed Grid.h to Lattice.h and tried to add "Lattice" to all functions there. 
remora: We make mostly use of ADL so instead of remora(row(m,a)) we write row(m,a)

further i moved the implementation of MOEAD to its own cpp file. Does not happen often that a class has no template parameters, but if it happens, we should put the big functions into the dll. Keeps the headers a litle bit nicer.

If you are happy i am going to merge.

**Comment by bjornbugge** [comment]
Looks good to me!

**Pull request merged**